File: native/swift/swift-export-standalone/testData/execution/classMappings/classMappings.kt
// KIND: STANDALONE
// FREE_COMPILER_ARGS: -opt-in=kotlin.native.internal.InternalForKotlinNative
// MODULE: ClassMappings
// FILE: classes.kt
import kotlin.native.internal.reflect.objCNameOrNull

class FinalClass {
    class NestedFinalClass
}
open class OpenClass
private class PrivateClass : OpenClass()

fun isAnyClassNameNull(): Boolean = Any::class.objCNameOrNull == null
fun getAnyClassName(): String = Any::class.objCNameOrNull!!

fun isFinalClassNameNull(): Boolean = FinalClass::class.objCNameOrNull == null
fun getFinalClassName(): String = FinalClass::class.objCNameOrNull!!

fun isNestedFinalClassNameNull(): Boolean = FinalClass.NestedFinalClass::class.objCNameOrNull == null
fun getNestedFinalClassName(): String = FinalClass.NestedFinalClass::class.objCNameOrNull!!

fun isOpenClassNameNull(): Boolean = OpenClass::class.objCNameOrNull == null
fun getOpenClassName(): String = OpenClass::class.objCNameOrNull!!

fun isPrivateClassNameNull(): Boolean = PrivateClass::class.objCNameOrNull == null
fun getPrivateClassName(): String = PrivateClass::class.objCNameOrNull!!

// FILE: classes_in_namespace.kt
package namespace

import kotlin.native.internal.reflect.objCNameOrNull

class NamespacedFinalClass

fun isNamespacedFinalClassNameNull(): Boolean = NamespacedFinalClass::class.objCNameOrNull == null
fun getNamespacedFinalClassName(): String = NamespacedFinalClass::class.objCNameOrNull!!

File: native/swift/swift-export-standalone/testData/execution/classMappings/classMappings.swift
import ClassMappings
import KotlinRuntime
import ObjectiveC

func testType<T: KotlinBase>(_ name: String, _ type: T.Type) throws {
    try assertSame(
        actual: objc_getClass(name) as! T.Type,
        expected: type
    )
}

func testAnyClass() throws {
    try testType(getAnyClassName()!, KotlinBase.self)
}

func testFinalClass() throws {
    try testType(getFinalClassName()!, FinalClass.self)
}

func testNestedFinalClass() throws {
    try testType(getNestedFinalClassName()!, FinalClass.NestedFinalClass.self)
}

func testNamespacedFinalClass() throws {
    try testType(namespace.getNamespacedFinalClassName()!, namespace.NamespacedFinalClass.self)
}

func testOpenClass() throws {
    try testType(getOpenClassName()!, OpenClass.self)
}

func testPrivateClass() throws {
    try assertNil(getPrivateClassName())
}

class ClassMappingsTests : TestProvider {
    var tests: [TestCase] = []

    init() {
        providers.append(self)
        tests = [
            TestCase(name: "testAnyClass", method: withAutorelease(testAnyClass)),
            TestCase(name: "testFinalClass", method: withAutorelease(testFinalClass)),
            TestCase(name: "testNestedFinalClass", method: withAutorelease(testNestedFinalClass)),
            TestCase(name: "testNamespacedFinalClass", method: withAutorelease(testNamespacedFinalClass)),
            TestCase(name: "testOpenClass", method: withAutorelease(testOpenClass)),
            TestCase(name: "testPrivateClass", method: withAutorelease(testPrivateClass)),
        ]
    }
}